Tasks/VsTest/vstest.ts

```diff
@@ -0,0 +1,480 @@
+                setRegistryKeyForParallelExecution(vsVersion);
+                return;
+            }
+        }
+    }
+}
+
+function resetRunInParallel() {
+    tl.warning("Install Visual Studio 2015 Update 1 or higher on your build agent machine to run the tests in parallel.");
+    runInParallel = false;
+}
+
+function locateVSVersion(): Q.Promise<number> {
+    var defer = Q.defer<number>();
+    var vsVersion = parseFloat(vsTestVersion);
+    if (!isNaN(vsVersion)) {
+        defer.resolve(vsVersion);
+        return defer.promise;
+    }
+
+    var regPath = "HKLM\\SOFTWARE\\Microsoft\\VisualStudio";
+    regedit.list(regPath).on('data', function(entry) {
+        if (entry && entry.data && entry.data.keys) {
+            var subkeys = entry.data.keys;
+            var versions = getFloatsFromStringArray(subkeys);
+            if (versions && versions.length > 0) {
+                versions.sort();
+                defer.resolve(parseFloat(versions[versions.length - 1]));
+                return defer.promise;
+            }
+        }
+        defer.resolve(null);
+    });
+    return defer.promise;
+}
+
+function getFloatsFromStringArray(inputArray: string[]): string[] {
+    var outputArray: string[] = []
+    var count;
+    if (inputArray) {
+        for (count = 0; count < inputArray.length; count++) {
+            var floatValue = parseFloat(inputArray[count]);
+            if (!isNaN(floatValue)) {
+                outputArray.push(inputArray[count]);
+            }
+        }
+    }
+    return outputArray;
+}
+
+function setRegistryKeyForParallelExecution(vsVersion: number) {
+    var regKey = "HKCU\\SOFTWARE\\Microsoft\\VisualStudio\\" + vsVersion.toFixed(1) + "_Config\\FeatureFlags\\TestingTools\\UnitTesting\\Taef";
+    regedit.createKey(regKey, function(err) {
+        if (!err) {
+            var values = {
+                [regKey]: {
+                    'Value': {
+                        value: '1',
+                        type: 'REG_DWORD'
+                    }
+                }
+            };
+            regedit.putValue(values, function(err) {
+                if (err) {
+                    tl.warning("Error occured while setting registry key, Error: " + err);
+                }
+            });
+        } else {
+            tl.warning("Error occured while creating registry key, Error: " + err);
+        }
+    });
+}
+
+function readFileContents(filePath: string, encoding: string): Q.Promise<string> {
+    var defer = Q.defer<string>();
+
+    fs.readFile(filePath, encoding, (err, data) => {
+        if (err) {
+            defer.reject(new Error('Could not read file (' + filePath + '): ' + err.message));
+        }
+        else {
+            defer.resolve(data);
+        }
+    });
+
+    return defer.promise;
+}
+
+function pathExistsAsFile(path: string) {
+    try {
+        return tl.stats(path).isFile();
+    }
+    catch (error) {
```

Comment:
You don't want to do this.  Use tl.exist and don't catch.  Then check is a file.

The key is to not catch and conclusively return false.  The reason is if it doesn't exist, tl.exist will return false (and not throw).  the isFile will also not throw (only call if tl.exist returns true).

Cases where it will throw are things like access denied (EPERM).  In all those cases, swallowing the exception and having the task continue on the premise that it doesn't exist just leads to hard to diagnose issues downstream.  In all those cases the task should just fail.  (one outer try catch - see ShellScript task).  And it will fail with a diagnosable error to the user - EPERM on path x